File: apps/staff/lib/terms-gate.ts
/**
 * The staff Terms gate's decision, on its own (#570, ADR 0067).
 *
 * ADR 0066 had the gate bind at the next sign-in. It never does: the Staff
 * Session's expiry slides on every authenticated request, so a daily user signs
 * in once and never again. The gate therefore binds where the person actually
 * is — on their next PAGE NAVIGATION — and the middleware asks this function,
 * per request, what to do about it.
 *
 * It is a pure function of a pathname and one flag for one reason: this is the
 * whole of the gate's logic on the client side, and the alternative way to test
 * it is a Playwright run. That is not available here — the e2e suite cannot be
 * green in one run (passcodes are rationed to 10 per IP per 15 minutes, and the
 * suite needs more sign-ins than that) and it drives a stale dev stack that
 * builds nothing. So the decision lives here, where a test can state a case in
 * one line, and middleware.ts keeps only the plumbing.
 *
 * NOTHING HERE REVOKES, RE-MINTS OR RE-PROVES ANYTHING. A diversion is a
 * redirect and no more: the session cookie is untouched, no passcode is spent,
 * and the person returns to where they were going the moment they accept.
 */

/** Where the interstitial lives. */
export const TERMS_GATE_PATH = "/terms";

/**
 * The query parameter carrying where the person was going when they were
 * stopped, so accepting returns them there rather than to the dashboard.
 */
export const RETURN_PARAM = "next";

export type TermsGateInput = {
  /** The pathname being navigated to. */
  pathname: string;
  /** Its query string including the leading "?", or "" when there is none. */
  search?: string;
  /**
   * `terms_outstanding` from the session route.
   *
   * NULL/UNDEFINED IS NOT FALSE. The backend leaves it null on every response
   * that did not ask, and on the one that asked and whose read failed — and a
   * failed consent read must not sign 25 people out or divert every navigation
   * into an interstitial that cannot be rendered. Only an explicit `true`
   * diverts.
   */
  termsOutstanding: boolean | null | undefined;
};

export type TermsGateDecision =
  | { kind: "allow" }
  /**
   * Divert this navigation. The pieces rather than a URL, because the caller
   * holds a NextURL to clone and this module must stay framework-free so a test
   * can run it under plain node.
   */
  | { kind: "interstitial"; pathname: string; search: string };

const ALLOW: TermsGateDecision = { kind: "allow" };

/**
 * What to do with one navigation.
 *
 * The order of the checks is the ruling:
 *
 * 1. `/api/` IS NEVER GATED. The gate is a navigation gate and nothing else, so
 *    a sale in progress commits and no in-flight mutation is ever refused
 *    because a legal edition rolled over at midnight. The middleware
 *    short-circuits these before it even resolves a session; the check is
 *    repeated here so the rule is testable as a rule rather than as a line's
 *    position in a file.
 * 2. The interstitial and the sign-in page pass, or the diversion is a loop.
 * 3. Everything else is diverted when — and only when — an acceptance is
 *    explicitly outstanding. `/operator` included, deliberately: the operator
 *    who published the edition meets their own interstitial, so the person who
 *    re-gated everybody has demonstrably read the text.
 */
export function decideTermsGate(input: TermsGateInput): TermsGateDecision {
  const { pathname } = input;

  if (pathname.startsWith("/api/")) {
    return ALLOW;
  }
  if (isPathWithin(pathname, TERMS_GATE_PATH) || isPathWithin(pathname, "/login")) {
    return ALLOW;
  }
  if (input.termsOutstanding !== true) {
    return ALLOW;
  }

  const target = `${pathname}${input.search ?? ""}`;
  return {
    kind: "interstitial",
    pathname: TERMS_GATE_PATH,
    search: `?${RETURN_PARAM}=${encodeURIComponent(target)}`,
  };
}

/**
 * Where an accepted interstitial sends the person: what they asked for, or the
 * app root.
 *
 * The parameter is attacker-controllable — it is in a URL somebody can be
 * handed — so it is treated as a claim rather than a destination. Only a
 * same-site absolute path survives: anything protocol-relative ("//evil.test")
 * or scheme-bearing would be an open redirect out of the staff app, and a
 * return INTO the gate would be a loop. `/api/` is refused too: those addresses
 * are for fetches, and landing a browser on one is never what the person was
 * doing.
 */
export function termsGateReturnPath(next: string | null | undefined): string {
  if (!next || !next.startsWith("/")) {
    return "/";
  }
  // "//host" and "/\host" are both read as protocol-relative by browsers.
  if (next.startsWith("//") || next.startsWith("/\\")) {
    return "/";
  }
  // A raw control character or space in a Location header is a header-splitting
  // attempt, not a path. Percent-encoded ones are ordinary and stay.
  if (/[\u0000-\u0020\u007f]/.test(next)) {
    return "/";
  }
  const pathname = next.split(/[?#]/, 1)[0];
  if (isPathWithin(pathname, TERMS_GATE_PATH) || pathname.startsWith("/api/")) {
    return "/";
  }
  return next;
}

/**
 * Whether the edition being shown asks the Adulthood Declaration (#587,
 * ADR 0069).
 *
 * THE LABEL'S PRESENCE IS THE WHOLE ANSWER, and that is the decision this
 * function exists to state once for both staff surfaces. The API omits
 * `adulthood_declaration_label` from the payload when the edition in effect
 * carries no `label-adulthood-declaration` artifact, and refuses to serve the
 * gate at all when the edition asks but cannot word the box — so "there are
 * words" and "there is a box" are the same fact, and neither surface may decide
 * it any other way. Drawing a mandatory contractual checkbox with nothing
 * written beside it is the one thing §3 forbids outright.
 *
 * Which means introducing the declaration is a PUBLISH and not a deploy: this
 * binary asks whenever the operator's edition asks, and never otherwise.
 */
export function asksAdulthoodDeclaration(
  adulthoodDeclarationLabel: string | null | undefined,
): boolean {
  return typeof adulthoodDeclarationLabel === "string" && adulthoodDeclarationLabel.trim() !== "";
}

/** The answers a staff terms surface holds when its submit button is drawn. */
export type TermsGateAnswers = {
  /** The label the API served for the second box, absent where it does not ask. */
  adulthoodDeclarationLabel?: string | null;
  /** The Terms box, always owed on these two surfaces. */
  termsAccepted: boolean;
  /** The Adulthood Declaration box, meaningful only where it was drawn. */
  adulthoodDeclared: boolean;
};

/**
 * Whether every owed required box is ticked — the enabled/disabled state of
 * both staff submit buttons, decided in one place (#587).
 *
 * THE API IS THE GUARANTEE AND THIS IS THE COURTESY. An unticked box is refused
 * by the backend with TERMS_ACCEPTANCE_REQUIRED or
 * ADULTHOOD_DECLARATION_REQUIRED whatever this returns, before anything is
 * written; a curl walks straight past a disabled button and gets the same
 * refusal. What this buys is that a person does not submit a form they cannot
 * possibly have finished.
 *
 * A box that was never drawn is never owed, so an edition that does not ask is
 * satisfied by the Terms box alone. That asymmetry is why the label is a
 * parameter here: the answer state alone cannot tell "unticked" from "never
 * shown", and treating the second as the first would disable the button on
 * every edition published before the Artifact existed.
 */
export function termsGateAnswersComplete(answers: TermsGateAnswers): boolean {
  if (!answers.termsAccepted) {
    return false;
  }
  if (asksAdulthoodDeclaration(answers.adulthoodDeclarationLabel) && !answers.adulthoodDeclared) {
    return false;
  }
  return true;
}

/** A path is within a prefix when it IS it, or is nested under it. */
function isPathWithin(pathname: string, prefix: string): boolean {
  return pathname === prefix || pathname.startsWith(`${prefix}/`);
}

File: backend/internal/identity/service/termsgate.go
package service

import (
	"context"
	"strings"

	"github.com/peter/ticket_pos/backend/internal/consent"
	"github.com/peter/ticket_pos/backend/internal/identity"
	"github.com/peter/ticket_pos/backend/internal/identity/repository"
	"github.com/peter/ticket_pos/backend/internal/platform"
)

// The staff gate on a LIVE session (#570, ADR 0067's amendment to ADR 0066).
//
// ADR 0066 had the gate bind at the next sign-in. In practice it never binds:
// the Staff Session's expiry slides on every authenticated request, so a daily
// user signs in once and never again, and "at the next sign-in" is unbounded
// rather than the fortnight it was taken to be. The gate therefore binds where
// the person actually is — on their next PAGE NAVIGATION — and does so within
// a minute of the edition's day arriving.
//
// What that costs, and what it deliberately does not:
//
//   - THE SESSION IS LEFT UNTOUCHED. Not revoked, not re-minted, and no
//     passcode is burned. Passcodes are rationed (10 per IP per 15 minutes,
//     sharing a platform ceiling with Sale Confirmations), so making somebody
//     re-prove an address they have already proven is making them spend
//     something they may not be able to get — during an event, for a document
//     they have not read yet.
//   - THE GATE NEVER BINDS ON `/api/`. It is a navigation gate and nothing
//     else: a sale in progress commits, and no in-flight mutation is refused
//     because a legal edition rolled over between the click and the write. The
//     staff app's middleware short-circuits `/api/` before it ever asks, and
//     the predicate below is not on any authenticated API path — which is also
//     why it is NOT computed in buildSessionView. A consent read that failed
//     there would 500 every staff API request, which is precisely the in-flight
//     refusal this rules out.
//   - PUBLISHING REVOKES NOTHING, in either population, and no control offers
//     to. Both session kinds are server-side rows and a DELETE would be easy;
//     it buys nothing from customers (already re-gated on a live session) and
//     from the staff it would force 25 passcodes through a limit of 10.
//
// The publishing operator is not exempt: `/operator` is checked against this
// gate too, so the person who re-gated everybody has demonstrably read the
// text they published.

// TermsOutstanding reports whether this email owes a Terms Acceptance — the
// gate predicate, and the one place it is spelled.
//
// MEMBERSHIP, NOT EQUALITY (#560): outstanding is "has accepted, as an
// organizer, NO edition that still satisfies the gate". A correction published
// under somebody's feet leaves them clear; a gating edition lifts the floor
// above every stored acceptance and re-gates everyone, with no code and no data
// migration.
//
// It reads the satisfying set ONLY, and never the current edition. Deciding
// who owes something is a question about the version rows; what to SHOW them is
// a second question, asked afterwards and only of the people who owe.
func (s *Service) TermsOutstanding(ctx context.Context, email string) (bool, error) {
	satisfying, err := s.termsVersions.SatisfyingTermsEditions(ctx)
	if err != nil {
		return false, err
	}
	accepted, err := s.repo.HasSatisfyingTermsAcceptance(ctx, email, satisfying.IDs())
	if err != nil {
		return false, err
	}
	return !accepted, nil
}

// StaffTermsGate is what a live Staff Session owes before its holder may see
// another page: the interstitial's box, or nil when they owe nothing.
//
// It is gateOnTerms verbatim — the same predicate, the same pinned edition, the
// same pinned label locale, the same fifteen-minute single-use token — because
// the two surfaces must ask the identical question and record the identical
// evidence. The only difference is what the token buys: at the sign-in door it
// buys the session that was withheld, and here it buys nothing, because the
// person already holds one.
//
// pageLocale is the language the staff app is rendered in for this reader. The
// label is served in it, floored at the prevailing text when this edition does
// not publish it (acceptanceLabel), and the locale ACTUALLY served comes back
// on the view so the link beside the box opens the document the words came
// from.
func (s *Service) StaffTermsGate(ctx context.Context, email string, pageLocale platform.Locale) (*TermsRequiredView, error) {
	return s.gateOnTerms(ctx, email, pageLocale, s.now())
}

// SessionTermsAcceptance is an interstitial being answered by somebody who is
// already signed in: the session doing the answering, the token that pinned
// what they were shown, the one box, and the circumstances to record.
type SessionTermsAcceptance struct {
	// SessionID is the live Staff Session, recorded as evidence and used for
	// nothing else. It is neither spent nor re-minted by this act.
	SessionID string
	// Email is the session's email, from the authenticated request and never
	// from the body.
	Email string
	// Token is the gate token from the interstitial that showed the box.
	Token string
	// TermsAcceptance is the required box.
	TermsAcceptance bool
	// Evidence is derived from the request by the handler, never from the body.
	Evidence consent.Evidence
}

// AcceptTermsOnSession records one Staff Terms Acceptance for somebody the
// navigation gate stopped, and returns them to what they were doing (#570).
//
// WHAT THIS DOES NOT DO IS THE POINT: it mints no session, spends no passcode,
// extends no authority and revokes nothing. The person was signed in before
// and is signed in after, with the same session row; all that changed is that
// the platform now holds their acceptance of the current edition.
//
// The edition and the language recorded are the ones PINNED TO THE TOKEN by
// the interstitial that rendered the box (#537's held-answer rule, #567's
// locale beside it). The request that ticks a box renders no text and knows
// neither.
//
// THE UNTICKED BOX IS JUDGED BEFORE THE TOKEN IS SPENT, which is the one place
// this departs from the sign-in door. There, spending first is deliberate: the
// token is a Proof of Email Ownership, and a refused submission must not be
// retriable against the same proof. Here the token proves nothing — the session
// is the credential — so burning it would only cost the person a page reload
// for having clicked the wrong thing, and the recovery from the sign-in door's
// version of that is a passcode this ticket exists to avoid spending.
func (s *Service) AcceptTermsOnSession(ctx context.Context, submission SessionTermsAcceptance) error {
	if !submission.TermsAcceptance {
		return identity.ErrTermsAcceptanceRequired()
	}

	pending, err := s.repo.ConsumePendingStaffTerms(ctx, submission.Token)
	if err != nil {
		return err
	}
	now := s.now()
	// Unknown, spent, expired, or issued to somebody else: one refusal,
	// deliberately indistinguishable. The email check is what stops a token
	// obtained for one address from being redeemed by another's session — it
	// would buy an acceptance in the wrong person's name, which is the only
	// thing a stolen gate token could ever be worth.
	if pending == nil || now.After(pending.ExpiresAt) ||
		!strings.EqualFold(pending.Email, submission.Email) {
		return identity.ErrPendingTermsInvalid()
	}

	return s.repo.InsertTermsAcceptance(ctx, repository.StaffTermsAcceptance{
		Email:           pending.Email,
		TermsVersionID:  pending.TermsVersionID,
		Capacity:        capacityOrganizer,
		AcceptedAt:      now,
		IP:              submission.Evidence.IP,
		UserAgent:       submission.Evidence.UserAgent,
		SessionID:       submission.SessionID,
		OriginURL:       submission.Evidence.OriginURL,
		PresentedLocale: pending.LabelLocale,
	})
}

// GetSessionWithTermsGate is GetSession plus the one fact the staff app's
// middleware needs to decide whether to divert this navigation to the
// interstitial (#570).
//
// It is a separate method, and the flag is a POINTER, because only this route
// asks. Every other caller of buildSessionView — SessionAuth on every staff API
// request, a verify, an organization switch — leaves it null, meaning "not
// asked" rather than "nothing owed". A plain false would be a field that lies
// on every response but one, and the middleware treats anything but an explicit
// true as no diversion.
//
// A FAILED GATE READ DOES NOT FAIL THE REQUEST, which is the opposite of what
// the sign-in door does with the same read, and deliberately. There, refusing
// costs a person a retry. Here, the session route's only failure mode is a 401
// to the middleware, which redirects to /login AND DELETES THE COOKIE — so a
// database hiccup would sign 25 people out and make them each spend a passcode,
// which is the exact harm this ticket forbids. So the flag is left null, the
// failure is logged, and this navigation proceeds; the next one asks again, and
// the acceptance is still owed at the sign-in door and at the interstitial,
// both of which fail closed.
func (s *Service) GetSessionWithTermsGate(ctx context.Context, sessionID string) (*SessionView, error) {
	view, err := s.GetSession(ctx, sessionID)
	if err != nil {
		return nil, err
	}

	outstanding, err := s.TermsOutstanding(ctx, view.Email)
	if err != nil {
		s.logger.Error("staff terms gate read", "error", err)
		return view, nil
	}
	view.TermsOutstanding = &outstanding
	return view, nil
}
